### ISSUE TITLE
Virtual Memory ACT

### ISSUE THREAD

**Issue by jjscheel** [riscv-admin/dev-partners#11]
### Technical Group

Privileged Spec IC

### ratification-pkg

Virtual Memory

### Technical Liaison

Dan Lustig, Andrea Mondelli

### Task Category

Arch Tests

### Task Sub Category

- [ ] gcc
- [ ] binutils
- [ ] gdb
- [ ] intrinsics
- [ ] Java
- [ ] KVM
- [ ] ld
- [ ] llvm
- [ ] Linux kernel
- [ ] QEMU
- [ ] Spike

### Ratification Target

3Q2023

### Statement of Work (SOW)

SOW: [link](https://docs.google.com/document/d/1FbpKkd0f15Zi2iz7iqRVy8vxasSX7qZ0H8oGnIwboi8/)

**SOW Signoffs:** (delete those not needed)
  - Task group liaison sign-off date:
  - Development partner sign-off date:
  - ACT SIG sign-off date (if ACT work):

### Waiver

- [X] Freeze
- [X] Ratification

**Comment by jjscheel** [comment]
Status from Umer:
> VM ACTs: Tests are getting developed under supervision of Allen. PR is expected in the next month hopefully.

**Comment by UmerShahidengr** [comment]
Update Apr 25th, 2023 => The Trap handler PR has been merged successfully. It has removed all known bugs related to VM and ePMP trap. The PR details are [here](https://github.com/riscv-non-isa/riscv-arch-test/pull/318)
Next week task => Complete macro definitions to setup page tables for SV39 and SV32, and start developing the tests according to the developed test plan

**Comment by jjscheel** [comment]
Update from Allen: 
> 1. Got the next version of the trap handler working and tested (With help for 10xe folks) and merged.
> It is now possible to write tests of virtual memory,  transition between protections mode easily, and write tests that can partition > permissions inside test code and data.
> 2. continue to partition and Test Format Spec into  ACT Developer Spec and ACT User Spec, and continue to get new chair up to speed
> 3. no
> 4. I don't have any dates....

**Comment by jjscheel** [comment]
My sense is that this work won't be done by next meeting on May 9.  So, let's try a target of May 23.  

@UmerShahidengr, thoughts?

**Comment by UmerShahidengr** [comment]
I couldn't understand what do you mean by "done" here. 
For this week's target, the macros for SV39 and SV32 have been updated, so there are no unresolved dependencies in writing ACTs for VM (for now). Some of the baseline tests are [available here](https://github.com/Abdulwadoodd/riscv-arch-test/tree/vm_tests/riscv-test-suite/rv32i_m/vm_tests), but they are not ready for the PR yet. We are testing it through Sail log, and not putting anything in the signature file yet. I hope these tests will be updated in this week and some of the new tests will also be added for SV32.
But this project is still a long way away from completion (if you are calling completion as "done") as there are so many test and corner cases to cover but we will try to do our first PR during the next week.

**Comment by UmerShahidengr** [comment]
Progress Update May 7th, 2023 => Still working on defining the macros for SV39 page table setup.
Next Week goal => Develop basic tests for SV32 (RV32) and make the first PR.

In regards to the question of @jjscheel , our initial deadline for completing this work was 2023Q2 but it will take more time. @Abdulwadoodd can you comment on any target deadline?

**Comment by UmerShahidengr** [comment]
Update May 23rd, 2023 => Not much update on this task yet.

**Comment by jjscheel** [comment]
Thanks, @UmerShahidengr.  Removed from today's agenda.  Touch base next meeting.

@Abdulwadoodd , do you have a target completion date for all work?

**Comment by UmerShahidengr** [comment]
Update ⇾ June 12th, 2023:
@Abdulwadoodd is busy in client's project, so we have been searching for a helping hand within 10xE who will help him in this project. Sa far, there is not much update in this.

**Comment by jjscheel** [comment]
Marking as "Blocked" due to current constraints identified.

**Comment by UmerShahidengr** [comment]
Update ⇾ July 11th, 2023
No update.

**Comment by jjscheel** [comment]
Any outlook for being able to make progress here?  Just wondering.

**Comment by UmerShahidengr** [comment]
Update ⇾ August 15th, 2023
This task was given to some undergraduate students as summer project, and those students have developed some good tests, draft tests are available [here](https://github.com/Masooma82/riscv-arch-test/tree/main/riscv-test-suite/rv32i_m/vm_test/src). I will optimize these tests and will make a PR before the next meeting. Review process will start from this week, so we can expect it to be merged in this quarter hopefully.

**Comment by UmerShahidengr** [comment]
Update ⇾ September 12th, 2023
The tests have been submitted in riscv-arch-test repo [here](https://github.com/riscv-non-isa/riscv-arch-test/pull/382).
Switched the progress status to review.

**Comment by jjscheel** [comment]
Thanks, @UmerShahidengr.  I've added the new PR to the first entry.

Question on the new PR, do we believe this is the final PR for the work?

**Comment by UmerShahidengr** [comment]
@jjscheel yes. This is the final PR. It contains all tests for SV32 mode.

**Comment by jjscheel** [comment]
As discussed today, more PRs are needed for Sv39, and Sv57 modes.

**Comment by UmerShahidengr** [comment]
Update ⇾ September 26th, 2023
Nobody has started reviewing SV32 tests yet. So, no update on this.
SV39 and SV48 PRs are expected before the next meeting.

**Comment by UmerShahidengr** [comment]
Update ⇾ October 10th, 2023
All PRs have been submitted, here are the details of all PRs:
SV32: https://github.com/riscv-non-isa/riscv-arch-test/pull/382
SV39: https://github.com/riscv-non-isa/riscv-arch-test/pull/392
SV48: https://github.com/riscv-non-isa/riscv-arch-test/pull/393

Just review is left now

**Comment by UmerShahidengr** [comment]
Update ⇾ October 24th, 2023
Coverpoint definitions are in process

**Comment by UmerShahidengr** [comment]
Update Nov 3rd, 2023:
* SV39 coverpoints are complete, its coverage report and coverpoints are available here: [PR at CTG](https://github.com/riscv-software-src/riscv-ctg/pull/89), [PR at arch-tests repo](https://github.com/riscv-non-isa/riscv-arch-test/pull/407)
* * SV39 coverpoints are complete, its coverage report and coverpoints are available here: [PR at CTG](https://github.com/riscv-software-src/riscv-ctg/pull/90), [PR at arch-tests repo](https://github.com/riscv-non-isa/riscv-arch-test/pull/393)

**Comment by UmerShahidengr** [comment]
Update ⇾ December 12th, 2023
The review process is going on smoothly.

**Comment by UmerShahidengr** [comment]
Update ⇾ January 30th, 2024
The review process is going on.

**Comment by jjscheel** [comment]
@UmerShahidengr, is this stalled in the community?  If so, is it all pieces or some?

@allenjbaum, FYI.

**Comment by UmerShahidengr** [comment]
It is not stalled in the community, it is stalled from our side. After converting the coverpoints of PMP, @MuhammadHammad001 will translate VM coverpoints according to the new template too.

**Comment by UmerShahidengr** [comment]
Update ⇾ April 2nd, 2024
Stalled from our side. Coverpoint definitions will be resumed once PMP is done

**Comment by jjscheel** [comment]
Thanks for the update @UmerShahidengr.  Please let us know when  @MuhammadHammad001 starts his work on coverpoints.

**Comment by UmerShahidengr** [comment]
Update April 30th, 2024:
Since pmp work is complete from our side, @MuhammadHammad001 has been working on sv32 coverpoints which will be delivered before the next meeting

**Comment by UmerShahidengr** [comment]
Update June 11th, 2024:
@MuhammadHammad001 has been working on this one in parallel with PMP.

**Comment by jjscheel** [comment]
@UmerShahidengr, what is the status of the sv32 coverpoints?  What remains to be completed?

**Comment by UmerShahidengr** [comment]
@jjscheel we have been targeting this quarter to complete these coverpoints, @MuhammadHammad001 has been working on this one.

**Comment by jjscheel** [comment]
Thanks.  Does "this quarter" mean by end of June (2 weeks) or in the next 90 days?

**Comment by UmerShahidengr** [comment]
Our plan was to conclude it in this quarter means by the end of June but I doubt it will take some more time

**Comment by UmerShahidengr** [comment]
Update June 25th, 2024
@MuhammadHammad001 has been working on this one, the work is in progress for sv32 coverpoints.

**Comment by UmerShahidengr** [comment]
Update July 23rd, 2024:
@MuhammadHammad001 will be unavailable for next few weeks due to medical emergency in his family. This work will be resumed after the next month.

**Comment by jjscheel** [comment]
I'm sorry to hear that.  Thanks for the update.

**Comment by UmerShahidengr** [comment]
Update October 29th, 2024:
The PR of SV32 will be merged in the next release meeting of SIG ARCH.
CC @allenjbaum

**Comment by UmerShahidengr** [comment]
PR of SV32 ACTs and coverpoints is still pending and awaiting final review by @allenjbaum

**Comment by MuhammadHammad001** [comment]
PR for SV-32 VM is merged.

**Comment by jjscheel** [comment]
Complete
